### ISSUE TITLE
calico role is unable to unblock openstack ports 

### ISSUE THREAD

**Issue by ipolyzos** [CiscoCloud/microservices-infrastructure#968]
Attempting to proceed with calico deployment in openstack i get the following error when try to unblock the neutron  ports. Could you please advise ?

```
failed: [mi-worker-003] => (item=mi-worker-003) => {"changed": false, "cmd": "/usr/local/bin/neutron_port_update.py \"192.168.0.0/16\" \"fa:16:3e:a6:94:7d\"", "delta": "0:00:20.117342", "end":
"2016-01-08 10:03:28.770773", "failed": true, "failed_when_result": true, "item": "mi-worker-003", "rc": 1, "start": "2016-01-08 10:03:08.653431", "stdout_lines": [], "warnings": []}
stderr: Traceback (most recent call last):
  File "/usr/local/bin/neutron_port_update.py", line 125, in <module>
    catalog = get_catalog()
  File "/usr/local/bin/neutron_port_update.py", line 56, in get_catalog
    r = requests.post(auth_url, headers=headers, data=json.dumps(payload))
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 108, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
failed: [mi-worker-003] => (item=mi-worker-002) => {"changed": false, "cmd": "/usr/local/bin/neutron_port_update.py \"192.168.0.0/16\" \"fa:16:3e:1d:c1:42\"", "delta": "0:00:20.117476", "end":
"2016-01-08 10:03:49.056929", "failed": true, "failed_when_result": true, "item": "mi-worker-002", "rc": 1, "start": "2016-01-08 10:03:28.939453", "stdout_lines": [], "warnings": []}
stderr: Traceback (most recent call last):
  File "/usr/local/bin/neutron_port_update.py", line 125, in <module>
    catalog = get_catalog()
  File "/usr/local/bin/neutron_port_update.py", line 56, in get_catalog
    r = requests.post(auth_url, headers=headers, data=json.dumps(payload))
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 108, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
failed: [mi-worker-003] => (item=mi-worker-001) => {"changed": false, "cmd": "/usr/local/bin/neutron_port_update.py \"192.168.0.0/16\" \"fa:16:3e:aa:6f:44\"", "delta": "0:00:20.104057", "end":
"2016-01-08 10:04:09.325214", "failed": true, "failed_when_result": true, "item": "mi-worker-001", "rc": 1, "start": "2016-01-08 10:03:49.221157", "stdout_lines": [], "warnings": []}
stderr: Traceback (most recent call last):
  File "/usr/local/bin/neutron_port_update.py", line 125, in <module>
    catalog = get_catalog()
  File "/usr/local/bin/neutron_port_update.py", line 56, in get_catalog
    r = requests.post(auth_url, headers=headers, data=json.dumps(payload))
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 108, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
failed: [mi-worker-003] => (item=mi-edge-01) => {"changed": false, "cmd": "/usr/local/bin/neutron_port_update.py \"192.168.0.0/16\" \"fa:16:3e:4d:1d:3c\"", "delta": "0:00:20.108533", "end": "20
16-01-08 10:04:29.581687", "failed": true, "failed_when_result": true, "item": "mi-edge-01", "rc": 1, "start": "2016-01-08 10:04:09.473154", "stdout_lines": [], "warnings": []}
stderr: Traceback (most recent call last):
  File "/usr/local/bin/neutron_port_update.py", line 125, in <module>
    catalog = get_catalog()
  File "/usr/local/bin/neutron_port_update.py", line 56, in get_catalog
    r = requests.post(auth_url, headers=headers, data=json.dumps(payload))
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 108, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
failed: [mi-worker-003] => (item=mi-edge-02) => {"changed": false, "cmd": "/usr/local/bin/neutron_port_update.py \"192.168.0.0/16\" \"fa:16:3e:b6:c6:50\"", "delta": "0:00:20.106809", "end": "20
16-01-08 10:04:49.842076", "failed": true, "failed_when_result": true, "item": "mi-edge-02", "rc": 1, "start": "2016-01-08 10:04:29.735267", "stdout_lines": [], "warnings": []}
stderr: Traceback (most recent call last):
  File "/usr/local/bin/neutron_port_update.py", line 125, in <module>
    catalog = get_catalog()
  File "/usr/local/bin/neutron_port_update.py", line 56, in get_catalog
    r = requests.post(auth_url, headers=headers, data=json.dumps(payload))
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 108, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
```

**Comment by stevendborrelli** [comment]
@Zogg can you take a look at this an advise?

**Comment by siddharthist** [comment]
It looks like the problem is [here](https://github.com/CiscoCloud/microservices-infrastructure/blob/master/roles/calico/files/neutron_port_update.py#L56), which is where the script attempts to reach the server [`OS_AUTH_URL`](https://github.com/CiscoCloud/microservices-infrastructure/blob/master/roles/calico/files/neutron_port_update.py#L27). Are all your environment variables set correctly?

**Comment by Zogg** [comment]
As @siddharthist mentioned, a shell from which ansible is run must have environment variables set for openstack:

```
  environment:
    OS_USERNAME: "{{ lookup('env','OS_USERNAME') }}"
    OS_PASSWORD: "{{ lookup('env','OS_PASSWORD') }}"
    OS_TENANT_NAME: "{{ lookup('env','OS_TENANT_NAME') }}"
    OS_AUTH_URL: "{{ lookup('env','OS_AUTH_URL') }}"
```

**Comment by ipolyzos** [comment]
Trying to further debug this issue attempted to manually set the environmental vars and run the scripts. It seems that the script is able to reach the AUTH_URL however fails in a later step trying to parse a returned value. Below is the output got

```
#> python neutron_port_update.py 192.168.0.0/16 XX:XX:XX:XX:XX:XX
Traceback (most recent call last):
  File "neutron_port_update.py", line 128, in <module>
    ports = list_ports(token, public_url)
  File "neutron_port_update.py", line 83, in list_ports
    r = requests.get(auth_url, headers=headers)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 68, in get
    return request('get', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 450, in request
    prep = self.prepare_request(req)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 381, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 304, in prepare
    self.prepare_url(url, params)
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 357, in prepare_url
    raise InvalidURL(*e.args)
requests.exceptions.InvalidURL: Failed to parse: neutron.XXXX.XXXX.XXXX.XXXX:9696v2.0
```

**Comment by loshkovskyi** [comment]
@ipolyzos, I think it's a problem with the [public_url](https://github.com/CiscoCloud/microservices-infrastructure/blob/master/roles/calico/files/neutron_port_update.py#L82) variable. Can you post your  `OS_AUTH_URL` environment variable (but editing out sensitive information)?

It might be a good idea to add some logging to this script. Alternatively, we could think about starting managing port resources with Terraform. Such functionality has been recently made available: http://www.terraform.io/docs/providers/openstack/r/networking_port_v2.html

**Comment by stevendborrelli** [comment]
+1 if we can get this in terraform.

**Comment by ipolyzos** [comment]
+1 for using terraform 

@loshkovskyi  the **OS_AUTH_URL** value is *https://keystone.XXX.XXX.XXX.net:5000/v2.0*

**Comment by loshkovskyi** [comment]
@ipolyzos, sorry for long delay in my reply. I think I was able to reproduce this problem. In some environments Neutron public URL in Keystone catalog can contain a trailing slash which may result in such issues with the script run. You can check the Neutron public URL with:
```
 openstack catalog show neutron
```
Do you mind testing my patch https://github.com/CiscoCloud/microservices-infrastructure/commit/a612b3deccdc6fc366629b1b6770b90c4581ee08 and see if it fixes this issue for you? With this patch it doesn't matter if the URL contains the trailing slash or not.

**Comment by mattsday** [comment]
I had this issue and tested your patch against OpenStack Mitaka.

It worked perfectly for me. In my case I didn't have a trailing '/':

```
ubuntu@mantl-bootstrap:~$  openstack catalog show neutron
+-----------+-------------------------------------+
| Field     | Value                               |
+-----------+-------------------------------------+
| endpoints | RegionOne                           |
|           |   internal: http://10.51.8.150:9696 |
|           | RegionOne                           |
|           |   public: http://10.51.8.150:9696   |
|           | RegionOne                           |
|           |   admin: http://10.51.8.150:9696    |
|           |                                     |
| id        | 7aafaba700a0482ead4e62c4508e7d39    |
| name      | neutron                             |
| type      | network                             |
+-----------+-------------------------------------+
```

But it was still the workaround!

**Comment by siddharthist** [comment]
This was closed in #1167. I'll make a new issue for tracking using terraform to unblock the ports.
